### PR TITLE
[SPARK-50031][SQL] Add the `TryParseUrl` expression

### PR DESCRIPTION
**Pull request by jovanm-db**
### What changes were proposed in this pull request?
This PR adds `try_parse_url` expression that sets `failOnError` to false by default.

### Why are the changes needed?
INVALID_URL contains suggested fix for turning off ANSI mode. Now that in Spark 4.0.0 we have moved to ANSI mode on by default, we want to keep suggestions of this kind to the minimum. There exist implementations of `try_*` functions which provide safe way to get behavior as for ANSI mode off and suggestions of this kind should be sufficient.

In this case, try expressions were missing so new expressions were added to patch up the missing implementations.

### Does this PR introduce _any_ user-facing change?
Yes, new expression added. 

### How was this patch tested?
Tests added. 

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala [hunk 1]
```diff
@@ -129,7 +129,7 @@ case class UrlDecode(child: Expression, failOnError: Boolean = true)
 case class TryUrlDecode(expr: Expression, replacement: Expression)
   extends RuntimeReplaceable with InheritAnalysisRules {
 
-  def this(expr: Expression) = this(expr, UrlDecode(expr, false))
+  def this(expr: Expression) = this(expr, UrlDecode(expr, failOnError = false))
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:132]: Avoid changes in other expressions unrelated to this PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala [hunk 2]
```diff
@@ -169,6 +169,36 @@ object ParseUrl {
...
+       query=1
+      > SELECT _FUNC_('http://spark.apache.org/path?query=1', 'QUERY', 'query');
+       1
+  """,
+  since = "4.0.0",
+  group = "url_funcs")
+case class TryParseUrl(params: Seq[Expression], replacement: Expression)
+  extends RuntimeReplaceable with InheritAnalysisRules {
+  def this(children: Seq[Expression]) = this(children, ParseUrl(children, failOnError = false))
+  def this() = this(Seq.empty)
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:191]: Why are we adding this constructor? ParseUrl needs to be called with children, so there is no need to add this constructor to try version.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 3]
```diff
@@ -362,12 +362,16 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase with ExecutionE
...
+      url: UTF8String,
+      e: URISyntaxException,
+      suggestedFunc: String)
+  : SparkIllegalArgumentException = {
     new SparkIllegalArgumentException(
       errorClass = "INVALID_URL",
       messageParameters = Map(
         "url" -> url.toString,
-        "ansiConfig" -> toSQLConf(SQLConf.ANSI_ENABLED.key)),
+        "functionName" -> suggestedFunc),
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:374]: Could we please just push try_parse_url to the error-conditions and not have this parameter. If the need comes that we have different functionNames we will update, but for now we want to push as much of the text as possible to error-conditions

**Review by mihailomilosevic2001 — COMMENTED** [review]
Great work Jovan, please address the comments I left and we can move this PR out of draft phase.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 4]
```diff
@@ -3207,7 +3207,7 @@
   },
   "INVALID_URL" : {
     "message" : [
-      "The url is invalid: <url>. If necessary set <ansiConfig> to \"false\" to bypass this error."
+      "The url is invalid: <url>. Use try_parse_url to tolerate invalid URL and return NULL instead."
```
Comment (mihailom-db) [inline common/utils/src/main/resources/error/error-conditions.json:3210]: Please add quotes to the function name, see some other errors. This is just to keep it in the same state.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 5]
```diff
@@ -362,12 +362,11 @@ private[sql] object QueryExecutionErrors extends QueryErrorsBase with ExecutionE
         "groupIndex" -> groupIndex.toString()))
   }
 
-  def invalidUrlError(url: UTF8String, e: URISyntaxException): SparkIllegalArgumentException = {
+  def invalidUrlError(url: UTF8String, e: URISyntaxException)
+  : SparkIllegalArgumentException = {
```
Comment (mihailom-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:366]: nit: Please fix this change in style, as it is not necessary.

**Review by mihailomilosevic2001 — APPROVED** [review]

File: common/utils/src/main/resources/error/error-conditions.json [hunk 6]
```diff
@@ -3207,7 +3207,7 @@
   },
   "INVALID_URL" : {
     "message" : [
-      "The url is invalid: <url>. If necessary set <ansiConfig> to \"false\" to bypass this error."
+      "The url is invalid: <url>. Use \"try_parse_url\" to tolerate invalid URL and return NULL instead."
```
Comment (mihailom-db) [inline common/utils/src/main/resources/error/error-conditions.json:3210]:
```suggestion
      "The url is invalid: <url>. Use `try_parse_url` to tolerate invalid URL and return NULL instead."
```

**Review by mihailomilosevic2001 — APPROVED** [review]
LGTM, just one more nit comment

**Review by srielau — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala [hunk 7]
```diff
@@ -169,6 +169,35 @@ object ParseUrl {
...
+@ExpressionDescription(
+  usage = "_FUNC_(url, partToExtract[, key]) - Extracts a part from a URL.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_('http://spark.apache.org/path?query=1', 'HOST');
+       spark.apache.org
+      > SELECT _FUNC_('http://spark.apache.org/path?query=1', 'QUERY');
+       query=1
+      > SELECT _FUNC_('http://spark.apache.org/path?query=1', 'QUERY', 'query');
+       1
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:184]: Can you show an example of the NULL result?

File: python/pyspark/sql/functions/builtin.py [hunk 8]
```diff
@@ -12912,6 +12912,108 @@ def substr(
         return _invoke_function_over_columns("substr", str, pos)
 
 
+@_try_remote_functions
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions/builtin.py:12915]: can we add a one unittest in, e.g., test_functions.py? then the tests will be resued in both Spark connect and classic.

**Review by HyukjinKwon — APPROVED** [review]

**Review by dongjoon-hyun — DISMISSED** [review]
Hi, @jovanm-db . 

IIUC, the PR title is misleading because the main contribution will be introducing a new built-in function `try_parse_url` at Apache Spark 4.0.0. `Improve error messages for ParseURL` looks like a side-effect to me.

The PR (and the commit) had better have one theme. I'd like to recommend to remove the change of `common/utils/src/main/resources/error/error-conditions.json` file from this PR and you can make another PR to handle it independently.

**Comment by dongjoon-hyun** [comment]
> @dongjoon-hyun I do agree that the title is not reflecting the right change, but opening a new PR might be an overkill and also cause confusion in development later. The main point of this PR is what ticket suggests `Remove the ANSI config suggestion in INVALID_URL`. I would suggest renaming the title, but keeping the changes in one PR, as these changes suggest the reason for removing and also the solution/substitute suggestion we provide for spark users. Separating them would make developer work much harder, and the second PR you suggested would be mostly tests changes.

No, I disagree with your opinion, "opening a new PR might be an overkill and also cause confusion in development later", @jovanm-db . Technically, `adding try_parse_url` and `recommending try_parse_url` is not the same at all. :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala [hunk 9]
```diff
@@ -169,6 +169,37 @@ object ParseUrl {
...
+  examples = """
+    Examples:
+      > SELECT _FUNC_('http://spark.apache.org/path?query=1', 'HOST');
+       spark.apache.org
+      > SELECT _FUNC_('http://spark.apache.org/path?query=1', 'QUERY');
+       query=1
+      > SELECT _FUNC_('inva lid://spark.apache.org/path?query=1', 'QUERY');
+       NULL
+      > SELECT _FUNC_('http://spark.apache.org/path?query=1', 'QUERY', 'query');
+       1
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:186]: Could you add an example which distiguish the new expression from `ParseUrl`, actually when it returns NULL.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala (hunk 9 above)
Comment (mihailom-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:186]: There is this test:
```> SELECT _FUNC_('inva lid://spark.apache.org/path?query=1', 'QUERY');
       NULL
```

File: python/pyspark/sql/functions/builtin.py [hunk 10]
```diff
@@ -13090,6 +13090,108 @@ def substr(
...
+def try_parse_url(
+    url: "ColumnOrName", partToExtract: "ColumnOrName", key: Optional["ColumnOrName"] = None
+) -> Column:
+    """
+    URL function: Extracts a specified part from a URL. If a key is provided,
```
Comment (mihailom-db) [inline python/pyspark/sql/functions/builtin.py:13098]: Could you update this description to give the similar message as other try functions?

File: python/pyspark/sql/tests/test_functions.py [hunk 11]
```diff
@@ -333,6 +333,14 @@ def test_rand_functions(self):
         rndn2 = df.select("key", F.randn(0)).collect()
         self.assertEqual(sorted(rndn1), sorted(rndn2))
 
+    def test_try_parse_url(self):
```
Comment (mihailom-db) [inline python/pyspark/sql/tests/test_functions.py:336]: Could we also add a case here that returns null? This is to make sure python works as well with invalid urls.

**Review by mihailomilosevic2001 — APPROVED** [review]
LGTM, please address the comments so we can merge this in

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala [hunk 12]
```diff
@@ -169,6 +169,37 @@ object ParseUrl {
   private val REGEXSUBFIX = "=([^&]*)"
 }
 
+/**
+ * Extracts a part from a URL
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:173]: This comment is useless. Let's remove it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala [hunk 13]
```diff
@@ -169,6 +169,37 @@ object ParseUrl {
...
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(url, partToExtract[, key]) - Extracts a part from a URL.",
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/urlExpressions.scala:176]: `ParseUrl` has the same description. Could write a few words what destiguish new expression from `ParseUrl`.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Review by mihailomilosevic2001 — APPROVED** [review]

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
@jovanm-db Please, update PR's description according the recent changes in the PR.

**Comment by MaxGekk** [comment]
@jovanm-db Congratulations with your first contribution to Apache Spark!

**Comment by dongjoon-hyun** [comment]
Thank you, @jovanm-db and @MaxGekk and all.

Congratulations to @jovanm-db .

**Pull request closed without merging**